flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -156,16 +175,47 @@ public void unregisterUnusedState(long lowestCheckpointID) {
                         subsumed.add(entry.stateHandle);
                     }
                     it.remove();
+                } else {
+                    markCheckpointInUseAction.accept(entry.createdByCheckpointID);
                 }
             }
         }
-
         LOG.trace("Discard {} state asynchronously", subsumed.size());
         for (StreamStateHandle handle : subsumed) {
             scheduleAsyncDelete(handle);
         }
     }
 
+    @Override
+    public void unregisterUnusedState(long lowestCheckpointID) {
+        unregisterState(lowestCheckpointID, (x) -> {});
```

Comment:
Now that subsumed checkpoints are "owned" by the registry, it has to discard them.

Reply:
This method is only used to delete unused states and called by `shutdown()`,  all checkpoints are discarded by `CompletedCheckpointStore` when shutdown. So,  I prefer not to discard checkpoints again.

Reply:
You are right, thanks for the clarification.